unchained-commons/src/main/java/unchained/commons/contract/Factory.java

```diff
@@ -0,0 +1,22 @@
+package unchained.commons.contract;
+
+import unchained.commons.utils.ReflectionUtils;
+
+import java.util.function.Function;
+
+@FunctionalInterface
+public interface Factory<T> extends Function<Object[], T> {
+
+    T create(Object... arguments);
```

Comment:
Passing in varargs arguments as the input to a factory encourages [the data clump anti-pattern](https://martinfowler.com/bliki/DataClump.html). We should instead have another interface such as `FactoryConfiguration` or `FactoryInstanceConfigurer` which can be passed to this method, and to generically bind the times at compile time we should have this be a parameter type for our factory:

```java
public interface Factory<T, C extends FactoryInstanceConfigurer<T, C>> {

    T create(C configuration);

    // ...

}
```

Reply:
I don't like the `Object[]` either. But adding another object, e.g. `FactoryConfiguration` is not in line with my actual intention which is making factories simple. If we decide to do it I would prefer to fall back to the class factory method without any reflection magic.